File: src/main/java/monad/Monad.java
package monad;

import common.DataClass;

import java.util.function.Function;

public interface Monad<T> {
    Monad<T> bind(Function<T, ? extends Monad<T>> f);
}

abstract class Maybe<T> extends DataClass implements Monad<T> {
}

class Just<T> extends Maybe<T> {

    private final T value;

    Just(T value) {
        this.value = value;
    }

    @Override
    public Monad<T> bind(Function<T, ? extends Monad<T>> f) {
        return f.apply(value);
    }
}

class Nothing<T> extends Maybe<T> {

    @Override
    public Monad<T> bind(Function<T, ? extends Monad<T>> f) {
        return new Nothing<>();
    }
}

File: src/main/kotlin/root_k/monad/monad.kt
package root_k.monad

import root_k.monad.Maybe.Just
import root_k.monad.Maybe.None

sealed class Maybe<out A> {
    object None : Maybe<Nothing>()
    data class Just<out A>(val value: A) : Maybe<A>()

    infix fun <B> bind(f: (A) -> Maybe<B>): Maybe<B> = when (this) {
        is None -> this
        is Just -> f(value)
    }
}

fun main(args: Array<String>) {
    None bind { x ->
        print("this will not be applied")
        None
    } // None

    Just(10) bind { Just(it / 2) } // Just 5
}